README.md
# shell-games

[![Circle CI](https://circleci.com/gh/GUI/lua-shell-games.svg?style=svg)](https://circleci.com/gh/GUI/lua-shell-games)

A Lua library to help execute shell commands more easily and safely.

- Easily execute shell commands, while capturing the command's output and exit code. Includes compatibility across versions of Lua, LuaJIT, and OpenResty where `io.popen` may not return exit codes (pre Lua 5.2 behavior).
- Utilities to quote and escape shell arguments for safer, less error-prone execution.

When executing shell commands, shell-games wraps either `os.execute` or `io.popen` (depending on whether the output is being captured).

*OpenResty Note:* If using shell-games with OpenResty, be aware that executing shell commands with this library is a blocking operation. This may or may not be okay depending on the nginx phase (blocking in `init` is okay) and your application's requirements. You may want to consider alternatives like [lua-resty-shell](https://github.com/juce/lua-resty-shell) or [lua-resty-exec](https://github.com/jprjr/lua-resty-exec) if blocking is a concern.

## Installation

Via [LuaRocks](https://luarocks.org):

```sh
luarocks install shell-games
```

Or via [OPM](https://opm.openresty.org):

```sh
opm get GUI/lua-shell-games
```

## Usage

```lua
local shell = require "shell-games"

-- Execute a command, with error handling.
local result, err = shell.run({ "touch", "/tmp/hello.txt" })
if err then
  print(err)
  exit(1)
end

-- Executue a command, capturing both its stdout and stderr output.
local result, err = shell.capture_combined({ "ls", "-l", "/tmp" })
print("Exit code: " .. result["status"])
print("Command output: " .. result["output"])

-- Quoting
shell.quote("ls") -- "ls"
shell.quote("It's happening.") -- [['It'"'"'s happening.']]
shell.quote("$PATH") -- "'$PATH'"

-- Quote and join
shell.join({ "ls", "-l", "/tmp/foo bar" }) -- "ls -l '/tmp/foo bar'"
```

## Compatibility

Tested against:

- Lua 5.1
- Lua 5.2
- Lua 5.3
- LuaJIT 2.0
- LuaJIT 2.0 compiled with `LUAJIT_ENABLE_LUA52COMPAT`
- LuaJIT 2.1-beta
- LuaJIT 2.1-beta compiled with `LUAJIT_ENABLE_LUA52COMPAT`
- OpenResty

## API

### `run`

**syntax:** `result, err = shell.run(command[, options])`

Execute a shell command.

The `comamnd` argument must be passed as a table of individual arguments that will be escaped. While this is the recommended way to execute commands, see [`run_raw`](#run_raw) if you need to execute an arbitrary shell command from a string.

The `options` table accepts the following fields:

- `capture` (boolean): Whether or not to capture the command's output. (default: `false`)

  By default, command output will not be captured, so any stdout/stderr generated by the command will be displayed in the parent process.

  If `true`, then the command's stdout will be captured and returned in the result, but stderr will still be printed in the parent process. To capture the command's stderr too, it must be redirected. See [`capture_combined`](#capture_combined) and [`capture`](#capture) for convenience wrappers for more easily running commands while capturing output.
- `chdir` (string): Change the current working directory to this path before executing the command. (default: `nil`)
- `env` (table): Set environment variables before executing the command. Accepts a table of environment variable names and values. (default: `{}`)
- `stderr` (string): Redirect the command's stderr output to a different path. (default: `nil`)
- `stdout` (string): Redirect the command's stdout output to a different path. (default: `nil`)
- `umask` (string): Change the process's umask before executing the command. (default: `nil`)

If executing the command fails (returning a non-0 exit code), then `err` will be a string describing the error. In the case of failure, the `result` table will still be returned (with the command's exit code and output reflected in the table).

The returned `result` table has the following fields:

- `command` (string): A string that shows the full command that was executed, after taking into account escaping and the options.
- `status` (integer): The exit code of the command.
- `output` (string): This field is only present if the `capture` option was enabled. If capturing was enabled, then this reflects the output of the command. By default, this will only contains the stdout from the command, and not the stderr. See [`capture_combined`](#capture_combined) and [`capture`](#capture) for convenience wrappers for more easily running commands while capturing output.

```lua
-- Basic example
local result, err = shell.run({ "ls", "-l", "/tmp" })

-- Example with options
local result, err = shell.run({ "ls", "-l", "/tmp" }, {
  capture = true,
  chdir = "/tmp",
  env = {
    LD_LIBRARY_PATH = "/usr/local/lib",
  },
  stderr = "/tmp/stderr.log",
  stdout = "/tmp/stdout.log",
  umask = "0022",
})
```

### `capture`

**syntax:** `result, err = shell.capture(command[, options])`

Execute a shell command, capturing stdout as a string.

This is a convenience wrapper around executing a command with [`run`](#run) while capturing stdout. It is equivalent to calling `run` with the options set to `{ capture = true }`.

The `command` and `options` arguments are given in the same format as in [`run`](#run).

```lua
local result, err = shell.capture({ "ls", "-l", "/tmp" })
```

### `capture_combined`

**syntax:** `result, err = shell.capture_combined(command[, options])`

Execute a shell command, capturing both stdout and stderr as a single string.

This is a convenience wrapper around executing a command with [`run`](#run) while capturing both stdout and stderr (by redirecting stderr to stdout). It is equivalent to calling `run` with the options set to `{ capture = true, stderr = "&1" }`.

The `command` and `options` arguments are given in the same format as in [`run`](#run).

```lua
local result, err = shell.capture_combined({ "ls", "-l", "/tmp/non-existent" })
```

### `run_raw`

**syntax:** `result, err = shell.run_raw(command[, options])`

Execute a shell command given as a raw string.

Usually, using [`run`](#run), [`capture`](#capture), or [`capture_combined`](#capture_combined) is recommended, which all accept the command to run as a table of individual arguments (since these handle shell escaping and quoting for you). However, in cases where you need to execute unescaped commands, this `run_raw` can be used to directly run a command string verbatim.

The `comamnd` argument must be passed as a string. Be sure to handle any escaping manually within this string.

The `options` argument is given in the same format as in [`run`](#run).

```lua
local result, err = shell.run_raw("echo $PATH", {
  capture = true,
})
print(result["output"])
```

### `quote`

**syntax:** `quoted_string = shell.quote(str)`

Return a shell-escaped version of the string. The escaped string can safely be used as one token in a shell command.

```lua
shell.quote("ls") -- "ls"
shell.quote("It's happening.") -- [['It'"'"'s happening.']]
shell.quote("$PATH") -- "'$PATH'"
```

### `join`

**syntax:** `quoted_string = shell.join(table)`

Accepts a table of individual command arguments, which will be escaped (using [`quote`](#quote)) and joined together by spaces. Suitable for turning a list of command arguments into a single command string.

```lua
shell.join({ "ls", "-l", "/tmp/foo bar" }) -- "ls -l '/tmp/foo bar'"
```

## Development

After checking out the repo, Docker can be used to run the test suite:

```sh
docker-compose run --rm app make test
```

### Release Process

To release a new version to LuaRocks and OPM:

- Ensure `CHANGELOG.md` is up to date.
- Update the `_VERSION` in `lib/shell-games.lua`.
- Move the rockspec file to the new version number (`git mv shell-games-X.X.X-1.rockspec shell-games-X.X.X-1.rockspec`), and update the `version` and `tag` variables in the rockspec file.
- Commit and tag the release (`git tag -a vX.X.X -m "Tagging vX.X.X" && git push origin vX.X.X`).
- Run `make release VERSION=X.X.X`.


CHANGELOG.md
# shell-games Change Log

## 1.1.0 - 2019-09-30

### Fixed
- Improve reliability for OpenResty versions prior to 1.15 by eliminating io.popen calls in these older versions (when capturing output, temporary files will be used instead).

## 1.0.2 - 2019-09-29

### Fixed
- Improve reliability when used in OpenResty if OpenResty 1.15 or newer is being used (which eliminates the need for nginx-related signal workarounds).

## 1.0.1 - 2019-05-01

### Fixed
- Fix the `env` option for setting environment variables.

## 1.0.0 - 2018-12-11

### Added
- Initial release.


spec/run_spec.lua
describe("run", function()
  local shell = require "shell-games"
  local clean_spec_tmp = require "spec/support/clean_spec_tmp"
  local table_keys = require "spec/support/table_keys"

  before_each(function()
    clean_spec_tmp()
  end)

  it("returns status, but no output by default", function()
    local result, err = shell.run({ "echo", "-n", "foo" })
    assert.are.same({
      command = "echo -n foo",
      status = 0,
    }, result)
    assert.are.equal(nil, err)
  end)

  it("capture option", function()
    local result, err = shell.run({ "echo", "-n", "foo" }, { capture = true })
    assert.are.same({
      command = "echo -n foo",
      status = 0,
      output = "foo",
    }, result)
    assert.are.equal(nil, err)

    result, err = shell.run({ "echo", "-n", "foo" }, { capture = false })
    assert.are.same({
      command = "echo -n foo",
      status = 0,
    }, result)
    assert.are.equal(nil, err)

    assert.has.error(function()
      shell.run({ "echo", "-n", "foo" }, { capture = "true" })
    end, "bad option 'capture' (boolean expected, got string)")
  end)

  it("chdir option", function()
    local file = io.open("spec/tmp/run chdir.txt", "w")
    file:write("")
    file:close()

    local result, err = shell.run({ "ls", "-1", "run chdir.txt" })
    assert.are.same({
      command = "ls -1 'run chdir.txt'",
      status = 2,
    }, result)
    assert.are.equal("Executing command failed (exit code 2): ls -1 'run chdir.txt'", err)

    result, err = shell.run({ "ls", "-1", "run chdir.txt" }, { chdir = "spec/tmp" })
    assert.are.same({
      command = "cd spec/tmp && ls -1 'run chdir.txt'",
      status = 0,
    }, result)
    assert.are.equal(nil, err)

    assert.has.error(function()
      shell.run({ "ls", "-1", "chdir.txt" }, { chdir = 1 })
    end, "bad option 'chdir' (string expected, got number)")
  end)

  it("env option", function()
    local result, err = shell.run({ "env" }, { capture = true })
    assert.are.same({ "command", "output", "status" }, table_keys(result))
    assert.are.equal("env", result["command"])
    assert.are.equal(0, result["status"])
    assert.are.match("PATH=/", result["output"])
    assert.are_not.match("SHELL_GAMES_FOO=foo", result["output"])
    assert.are.equal(nil, err)

    result, err = shell.run({ "env" }, { capture = true, env = { SHELL_GAMES_FOO = "foo bar" } })
    assert.are.same({ "command", "output", "status" }, table_keys(result))
    assert.are.equal("env 'SHELL_GAMES_FOO=foo bar' env", result["command"])
    assert.are.equal(0, result["status"])
    assert.are.match("PATH=/", result["output"])
    assert.are.match("SHELL_GAMES_FOO=foo bar", result["output"])
    assert.are.equal(nil, err)

    result, err = shell.run({ "echo" }, { capture = true, env = { SHELL_GAMES_FOO = "foo bar" } })
    assert.are.same({ "command", "output", "status" }, table_keys(result))
    assert.are.equal("env 'SHELL_GAMES_FOO=foo bar' echo", result["command"])
    assert.are.equal(0, result["status"])
    assert.are.equal("\n", result["output"])
    assert.are.equal(nil, err)

    assert.has.error(function()
      shell.run({ "env" }, { env = "FOO=bar" })
    end, "bad option 'env' (table expected, got string)")
  end)

  it("stderr option", function()
    local result, err = shell.run({ "spec/support/generate-stdout-stderr" }, { stderr = "spec/tmp/run stderr.txt" })
    assert.are.same({
      command = "spec/support/generate-stdout-stderr 2> 'spec/tmp/run stderr.txt'",
      status = 0,
    }, result)
    assert.are.equal(nil, err)

    local file = io.open("spec/tmp/run stderr.txt")
    local file_output = file:read("*a")
    file:close()
    assert.are.equal("2. stderr\n4. stderr\n", file_output)

    assert.has.error(function()
      shell.run({ "spec/support/generate-stdout-stderr" }, { stderr = 1 })
    end, "bad option 'stderr' (string expected, got number)")
  end)

  it("stdout option", function()
    local result, err = shell.run({ "spec/support/generate-stdout-stderr" }, { stdout = "spec/tmp/run stdout.txt" })
    assert.are.same({
      command = "spec/support/generate-stdout-stderr 1> 'spec/tmp/run stdout.txt'",
      status = 0,
    }, result)
    assert.are.equal(nil, err)

    local file = io.open("spec/tmp/run stdout.txt")
    local file_output = file:read("*a")
    file:close()
    assert.are.equal("1. stdout\n3. stdout\n", file_output)

    assert.has.error(function()
      shell.run({ "spec/support/generate-stdout-stderr" }, { stdout = 1 })
    end, "bad option 'stdout' (string expected, got number)")
  end)

  it("stdout redirect to stderr", function()
    local result, err = shell.run({ "spec/support/generate-stdout-stderr" }, { stdout = "&2", stderr = "spec/tmp/run stdout-to-stderr.txt" })
    assert.are.same({
      command = "spec/support/generate-stdout-stderr 2> 'spec/tmp/run stdout-to-stderr.txt' 1>&2",
      status = 0,
    }, result)
    assert.are.equal(nil, err)

    local file = io.open("spec/tmp/run stdout-to-stderr.txt")
    local file_output = file:read("*a")
    file:close()
    assert.are.equal("1. stdout\n2. stderr\n3. stdout\n4. stderr\n", file_output)
  end)

  it("stderr redirect to stdout", function()
    local result, err = shell.run({ "spec/support/generate-stdout-stderr" }, { stderr = "&1", stdout = "spec/tmp/run stderr-to-stdout.txt" })
    assert.are.same({
      command = "spec/support/generate-stdout-stderr 2>&1 1> 'spec/tmp/run stderr-to-stdout.txt'",
      status = 0,
    }, result)
    assert.are.equal(nil, err)

    local file = io.open("spec/tmp/run stderr-to-stdout.txt")
    local file_output = file:read("*a")
    file:close()
    assert.are.equal("1. stdout\n3. stdout\n", file_output)
  end)

  it("umask option", function()
    local result, err = shell.run({ "umask" }, { capture = true })
    assert.are.same({
      command = "umask",
      status = 0,
      output = "0022\n",
    }, result)
    assert.are.equal(nil, err)

    result, err = shell.run({ "umask" }, { capture = true, umask = "077" })
    assert.are.same({
      command = "umask 077 && umask",
      status = 0,
      output = "0077\n",
    }, result)
    assert.are.equal(nil, err)

    assert.has.error(function()
      shell.run({ "umask" }, { umask = 077 })
    end, "bad option 'umask' (string expected, got number)")
  end)

  it("multiple options", function()
    local result, err = shell.run({ "echo", "foo" }, {
      env = { FOO = "bar" },
      chdir = "spec/tmp",
      umask = "077",
    })
    assert.are.same({
      command = "cd spec/tmp && umask 077 && env FOO=bar echo foo",
      status = 0,
    }, result)
    assert.are.equal(nil, err)
  end)

  it("quotes arguments", function()
    local result, err = shell.run({ "echo", "-n", "$PATH" }, { capture = true })
    assert.are.same({
      command = "echo -n '$PATH'",
      status = 0,
      output = "$PATH",
    }, result)
    assert.are.equal(nil, err)
  end)

  it("returns command errors", function()
    local result, err = shell.run({ "exit", "33" })
    assert.are.same({
      command = "exit 33",
      status = 33,
    }, result)
    assert.are.equal("Executing command failed (exit code 33): exit 33", err)
  end)

  it("raises error when table not given for args", function()
    assert.has.error(function()
      shell.run(nil)
    end, "bad argument #1 (table expected, got nil)")

    assert.has.error(function()
      shell.run("ls")
    end, "bad argument #1 (table expected, got string)")
  end)

  it("raises error when table not given for options", function()
    assert.has.error(function()
      shell.run({ "ls" }, "options")
    end, "bad argument #2 (table expected, got string)")
  end)

  it("raises error for unknown option", function()
    assert.has.error(function()
      shell.run({ "ls" }, { foobar = true })
    end, "bad option 'foobar' (unknown option)")
  end)
end)


spec/join_spec.lua
describe("join", function()
  local shell = require "shell-games"

  it("joins safe arguments", function()
    assert.are.equal("ls -l /tmp", shell.join({ "ls", "-l", "/tmp" }))
  end)

  it("joins and quotes unsafe arguments", function()
    assert.are.equal("ls -l '/tmp/foo bar'", shell.join({ "ls", "-l", "/tmp/foo bar" }))
  end)

  it("returns empty table as empty string", function()
    assert.are.equal("", shell.join({}))
  end)

  it("raises error when table not given", function()
    assert.has.error(function()
      shell.join(nil)
    end, "bad argument #1 (table expected, got nil)")

    assert.has.error(function()
      shell.join("test")
    end, "bad argument #1 (table expected, got string)")
  end)
end)


spec/quote_spec.lua
-- Quoting implementation and tests based on Python's shlex:
-- https://github.com/python/cpython/blob/v3.7.1/Lib/test/test_shlex.py#L296-L309
describe("quote", function()
  local shell = require "shell-games"

  it("quotes empty string", function()
    assert.are.equal("''", shell.quote(""))
  end)

  it("quotes nil as empty string", function()
    assert.are.equal("''", shell.quote(nil))
  end)

  it("returns false as string", function()
    assert.are.equal("false", shell.quote(false))
  end)

  it("returns number as string", function()
    assert.are.equal("3", shell.quote(3))
  end)

  it("returns other object types as strings", function()
    assert.are.match("'table: 0x%w+'", shell.quote({ foo = "bar" }))
  end)

  it("returns safe strings as-is", function()
    local safe = {
      "foobar",
      [[abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789@%_-+=:,./]],
    }
    for _, str in ipairs(safe) do
      assert.are.equal(str, shell.quote(str))
    end
  end)

  it("quotes string with spaces", function()
    assert.are.equal("'test file name'", shell.quote("test file name"))
  end)

  it("quotes string with double quotes", function()
    assert.are.equal([['"foo bar"']], shell.quote([["foo bar"]]))
  end)

  it("quotes unsafe strings", function()
    local unsafe = {
      [["]],
      [[`]],
      [[$]],
      [[\]],
      [[!]],
      "\233",
      "\224",
      "\223",
      [[&]],
      [[;]],
      [[{]],
      [[}]],
    }
    for _, str in ipairs(unsafe) do
      assert.are.equal(string.format([['test%sname']], str), shell.quote("test" .. str .. "name"))
      assert.are.equal(string.format([['test%s'"'"'name'"'"'']], str), shell.quote("test" .. str .. "'name'"))
    end
  end)
end)


lib/shell-games.lua
local _M = {
  _VERSION = "1.1.0"
}

-- Valid options for the `run` commands.
local RUN_OPTIONS = {
  capture = "boolean",
  chdir = "string",
  env = "table",
  stderr = "string",
  stdout = "string",
  umask = "string",
}

-- Detect whether the app is running in Lua 5.2 mode (including LuaJIT compiled
-- with the LUAJIT_ENABLE_LUA52COMPAT option), since that affects the
-- os.execute and pipe:close return codes.
--
-- Based on https://github.com/keplerproject/lua-compat-5.3/blob/v0.7/compat53/init.lua#L88-L91
local lua_version = _VERSION:sub(-3)
local is_luajit = (string.dump(function() end) or ""):sub(1, 3) == "\027LJ"
local is_luajit52 = is_luajit and #setmetatable({}, { __len = function() return 1 end }) == 1
local LUA52_MODE = lua_version > "5.1" or is_luajit52

-- Detect when the OpenResty version is < v1.15 (ngx_lua < v0.10.14). For these
-- older versions, io.popen is unstable due to nginx's SIGCHLD handler. So for
-- these versions, only ever use os.execute and capture output by writing it to
-- a temporary file.
--
-- OpenResty 1.15+ (ngx_lua 0.10.14+) should behave properly as long as the
-- "lua_sa_restart" option is left enabled (the default).
--
-- https://github.com/openresty/lua-nginx-module/issues/779
-- https://github.com/openresty/resty-cli/issues/35
-- https://github.com/openresty/lua-nginx-module/pull/1296
local NGX_UNSTABLE_POPEN = (ngx and ngx.config.ngx_lua_version < 10014) -- luacheck: globals ngx

-- Characters that need shell escaping.
local UNSAFE_CHARS = "[^A-Za-z0-9_@%%+=:,./-]"

local function assert_arg(index, value, expected_type)
  local actual_type = type(value)
  if actual_type ~= expected_type then
    error(string.format("bad argument #%d (%s expected, got %s)", index, expected_type, actual_type), 2)
  end
end

local function assert_run_options(options)
  for key, _ in pairs(options) do
    local expected_type = RUN_OPTIONS[key]
    if not expected_type then
      error(string.format("bad option '%s' (unknown option)", key), 2)
    else
      local actual_type = type(options[key])
      if actual_type ~= expected_type then
        error(string.format("bad option '%s' (%s expected, got %s)", key, expected_type, actual_type), 2)
      end
    end
  end
end

-- Alter the command line string to execute, depending on the run options
-- passed in. This allows for changing the command in common ways (eg, to set
-- environment variables, change the directory, or redirect output).
local function add_command_options(command, options)
  if options["env"] ~= nil then
    local env = { "env" }
    for env_key, env_value in pairs(options["env"]) do
      table.insert(env, env_key .. "=" .. env_value)
    end
    command = _M.join(env) .. " " .. command
  end

  if options["umask"] ~= nil then
    command = _M.join({ "umask", options["umask"] }) .. " && " .. command
  end

  if options["chdir"] ~= nil then
    command = _M.join({ "cd", options["chdir"] }) .. " && " .. command
  end

  if options["stderr"] ~= nil or options["stdout"] ~= nil then
    -- If we're attempting to redirect stderr/stdout and the command consists
    -- of multiple commands (either separated by "&&" or ";"), then wrap the
    -- entire command in a sub-shell. This is needed so that we capture the
    -- entire output, rather than just the output for the last command.
    if string.match(command, "&&") or string.match(command, ";") then
      if not string.match(command, "^sh -c ") then
        command = "sh -c " .. _M.quote(command)
      end
    end

    -- Redirect stderr. If a path is given, be sure to escape it, but if a
    -- we're just redirecting to another descriptor (eg, stdout, &1), then
    -- don't escape.
    if options["stderr"] ~= nil then
      if string.match(options["stderr"], "^&%d+$") then
        command = command .. " 2>" .. options["stderr"]
      else
        command = command .. " 2> " .. _M.quote(options["stderr"])
      end
    end

    if options["stdout"] ~= nil then
      if string.match(options["stdout"], "^&%d+$") then
        command = command .. " 1>" .. options["stdout"]
      else
        command = command .. " 1> " .. _M.quote(options["stdout"])
      end
    end
  end

  return command
end

local function execute(command)
  local result

  if LUA52_MODE then
    local _, _, status = os.execute(command)
    result = {
      status = status,
    }
  else
    -- os.execute's return signature is a bit different under Lua 5.1 or older.
    -- The exit code value is also shifted by one byte in Linux, so it needs to
    -- be adjusted:
    -- http://lua-users.org/lists/lua-l/2004-10/msg00109.html
    -- https://github.com/keplerproject/lua-compat-5.3/pull/17
    -- https://github.com/keplerproject/lua-compat-5.3/blob/v0.7/compat53/module.lua#L604-L606
    local status = os.execute(command)
    if string.sub(package.config, 1, 1) == "/" then
      status = status / 256
    end
    result = {
      status = status,
    }
  end

  return result
end

local function capture(command)
  local tmp_output_path

  -- By default we'll try to get the exit code from io.popen's handle:close()
  -- behavior. However, this functionality is only available in Lua 5.2+. So
  -- for Lua 5.1 and older, fallback to a workaround approach which consists of
  -- appending the status code to the shell output, and then parsing that out
  -- (http://lua-users.org/lists/lua-l/2009-06/msg00133.html).
  local status_code_workaround = false
  if not LUA52_MODE and not NGX_UNSTABLE_POPEN then
    status_code_workaround = true
  end

  if status_code_workaround or NGX_UNSTABLE_POPEN then
    -- Ensure that the command is wrapped in sub-shell, so we should always
    -- output the status code output afterwards, even if the underlying command
    -- exits early. This sub shell also ensures we properly handle the output
    -- redirection when outputting to a temp file.
    if not string.match(command, "^sh -c ") then
      command = "sh -c " .. _M.quote(command)
    end

    if status_code_workaround then
      command = command .. [[; echo -n "=====LUA_SHELL_GAME_STATUS_CODE:$?"]]
    elseif NGX_UNSTABLE_POPEN then
      -- For OpenResty < 1.15, capture output to a file to deal with unstable
      -- "popen" issues.
      tmp_output_path = os.tmpname()
      command = command .. " > " .. _M.quote(tmp_output_path)
    end
  end

  local result = {}
  local err
  local handle
  if NGX_UNSTABLE_POPEN then
    result, err = execute(command)
    handle = io.open(tmp_output_path, "r")
  else
    handle = io.popen(command, "r")
  end
  local all_output = handle:read("*a")

  if not status_code_workaround then
    if NGX_UNSTABLE_POPEN then
      os.remove(tmp_output_path)
    else
      local _, _, close_status = handle:close()
      result["status"] = close_status
    end

    result["output"] = all_output
  else
    handle:close()

    -- Parse the status code out of the output.
    if all_output then
      local match_output, match_status = string.match(all_output, [[^(.*)=====LUA_SHELL_GAME_STATUS_CODE:(%d+)$]])
      if match_output and match_status then
        result["output"] = match_output
        result["status"] = tonumber(match_status)
      end
    end

    if result["status"] == nil then
      -- This means we never got the "STATUS_CODE" output, so the entire
      -- sub-processes must have gotten killed off.
      err = "Command exited prematurely: " .. command .. "\nOutput: " .. (all_output or "")
    end
  end

  return result, err
end

-- Quoting implementation based on Python's shlex:
-- https://github.com/python/cpython/blob/v3.7.1/Lib/shlex.py#L310-L319
function _M.quote(str)
  local quoted_str = tostring(str)
  if str == nil or quoted_str == nil or #quoted_str == 0 then
    return "''"
  end

  if not string.match(quoted_str, UNSAFE_CHARS) then
    return quoted_str
  end

  return "'" .. string.gsub(quoted_str, "'", "'\"'\"'") .. "'"
end

function _M.join(args)
  assert_arg(1, args, "table")

  local quoted = {}
  for _, str in ipairs(args) do
    table.insert(quoted, _M.quote(str))
  end

  return table.concat(quoted, " ")
end

function _M.run_raw(command, options)
  assert_arg(1, command, "string")
  if options ~= nil then
    assert_arg(2, options, "table")
  else
    options = {}
  end
  assert_run_options(options)

  command = add_command_options(command, options)

  local result, err
  if options["capture"] then
    result, err = capture(command)
  else
    result, err = execute(command)
  end
  result["command"] = command

  if err == nil and result["status"] ~= 0 then
    err = "Executing command failed (exit code " .. (result["status"] or "") .. "): " .. result["command"]
    if options["capture"] then
      err = err .. "\nOutput: " .. (result["output"] or "")
    end
  end

  return result, err
end

function _M.run(args, options)
  assert_arg(1, args, "table")
  return _M.run_raw(_M.join(args), options)
end

function _M.capture(args, options)
  assert_arg(1, args, "table")
  if options ~= nil then
    assert_arg(2, options, "table")
    if options["capture"] ~= nil then
      error("bad option 'capture' (unknown option)")
    end
  else
    options = {}
  end
  options["capture"] = true

  return _M.run_raw(_M.join(args), options)
end

function _M.capture_combined(args, options)
  assert_arg(1, args, "table")
  if options ~= nil then
    assert_arg(2, options, "table")
    if options["stderr"] ~= nil then
      error("bad option 'stderr' (unknown option)")
    end
  else
    options = {}
  end
  options["stderr"] = "&1"

  return _M.capture(args, options)
end

return _M
